docs/torchelastic-operator-design.md

```diff
@@ -0,0 +1,169 @@
+## Motivation
+
+With job fault tolerance and elastic training, we can unlock a lot of features.
+
+Users can enable job priority and preemption in the cluster. Losing a task becomes acceptable and the user won't lose the entire job progress. More importantly, it will help guarantee SLAs of crititcal jobs, even in a cluster under resource pressure.
+
+Cost and GPU utilization will be further optimized with this feature, since users can launch jobs with partial resources and spot GPU instances can be used as well without worrying.
+
+## User Experience
+
+* Users should define the `min` and `max` number of tasks of a job instead of a fixed number. The TorchElastic Operator will launch jobs in Kubernetes, setup the needed network topology and manage the job lifecycle.
+* Users have the option to set the etcd endpoint used as the RDZV service for task coordination.
+* The desired `size`, being number of tasks, is an optional field in the range from `min` to `max`.
+* Users can easily create/delete a torch elastic job using `kubectl` using a job manifest.
+* Users are able to describe custom resources to monitor the job status.
+
+## High Level Design
+
+Torch elastic jobs does not have a concept of a parameter server, instead rank 0 (one of the workers) will become the leader. We can roughly think all the workers are equivalent and their communication is peer to peer.
+In this case, every pod should be able to talk with every other pod, and we need to create a `headless` service for every pod.
+
+A config with kind `TorchElasticJob` defines the job spec and the controller will reconcile against this definition. It will create/update/delete pods and services if there are any changes in the job orin the kubernetes resources (pods, services) changes owned by `TorchElasticJob`.
+
+```
+apiVersion: "kubeflow.org/v1alpha1"
+kind: "TorchElasticJob"
+metadata:
+  name: "classy-vision-job"
+spec:
+  rdzvEndpoint: "etcd-service:2379"
+  size: 3
+  min: 2
+  max: 5
+  spec:
+    containers:
+    - name: torchelasticworker
+      image: torchelastic/examples:0.1.0rc1
+      imagePullPolicy: Always
+      args:
+        - "s3://code_path/petctl/user/my_job/main.py"
+        - --config_file
+        - "/data/classy_vision/resnet50_synthetic_image_classy_config.json"
+        - "--checkpoint_folder"
+        - "/data/classy_vision/checkpoint""
+
+```
+
+*Network Communication*
+
+In this case, every pod should be able to talk with each other and we need to create headless service for every pod since they use hostname registered in rdzv endpoint to find peers.
+
+*Failure condition*
+
+Torch Elastic controller will only fail a job if active workers is under min size user specified. Otherwise, it will try to reschedule failed pods and maintain the desired task size.
+
+*Optional fields - rdzvEndpoint*
+
+Controller will create an etcd instance in Kubernetes for the job is `rdzvEndpoint` is not set by user. It could be high available etcd quorum or single etcd pods.
```

Comment:
"job is `rdzvEndpoint` is" -> "job **if** `rdzvEndpoint` is"

Reply:
I will make the change.